### PR TITLE
DFC- 63 - DFC | GA4 | Sandbox | Add journey guard to test pages

### PR DESCRIPTION
**Pull request by SophiaDWokoma**
### Description
I have used express session to implement the journey guard.

- Adds middleware function that that checks if user is on the home page and does not have an active session, and sets a session value if true, if user is not on the home page and does not have a session value then they are redirected to Journey Guard Page which has a link which the user can click to take them to Home Page where session is set and they are able to navigate freely.
- Adds middleware.js file for this function
- Uses middleware function in app.js file
- Changes path for home page from `/ `to` /welcome` to fix bug I was facing .

### Tickets
(DFC-63)[https://govukverify.atlassian.net/browse/DFC-63]

### Steps to Reproduce
To test:

1. Run App
2. Enter url of form page e.g. localhost:3000/organisation-type
3. Journey Guard Page should display
4. Navigate to Home page( Session ID is set/check dev tools)
5. Click the "Start Now" button.
6. Organisation type form should now display as session value has been set.

### CODE REVIEW

File: src/views/home.njk [hunk 1]
```diff
@@ -58,6 +58,10 @@ service." }) }}
     </div>
 
     <script>
-        localStorage.clear();
+        document.addEventListener("DOMContentLoaded", function () {
```
[the remark this answers is not in the document]
Comment (di-fabs) [reply src/views/home.njk:61]: So the solution needs to be host in the backend side. But what can be a good solution?
Start from the functional requirement: "redirect a user to the homepage, if not the user is starting on any page that is not the homepage", and try to apply it to our demo app (with the difficulty of our demo app - no backend).
A solution is to use a session you can store in the cookie. Account repo is doing that (https://signin.account.gov.uk/sign-in-or-create).
Example of possible (maybe) implementation: https://medium.com/@ethantcollins98/setting-and-using-cookies-with-a-node-js-express-server-49479673d043

**Review by di-fabs — CHANGES_REQUESTED** [review]

File: src/views/organisationType.njk [hunk 2]
```diff
@@ -51,17 +52,36 @@
...
+        if (storedOrganisationType) {
+          Array
+            .from(form.elements.organisationType)
+            .find((radio) => radio.value === storedOrganisationType)
+            .checked = true
+        }
+        form.style.display = 'block';
+      } else {
+        document
+          .getElementById('sorryMessage')
```
Comment (Joanneyoung01) [inline src/views/organisationType.njk:82]: So just looking at the patterns of the other pods, and although this is a good solution we want it to be scalable i think it'd be best that we create a nunjucks file the same way the accounts pod have then render that page if the error response comes back as unauthorised

Comment (Joanneyoung01) [reply src/views/organisationType.njk:82]: https://github.com/alphagov/di-account-management-frontend/blob/main/src/components/common/errors/404.njk here's an example :D

Comment (SophiaDWokoma) [reply src/views/organisationType.njk:82]: agreed, I initially created it as a file/separate page but then wasn't sure how to render that using the localStorage method so I deleted it :(

File: src/app.js [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
 const path = require("path");
+const session = require("express-session");
 const { configureNunjucks } = require("./config/nunjucks");
 const validateForm = require("./validateForm");
 const {
   setGa4ContainerId,
   setStatusCode,
   setTaxonomyValues,
   setPageTitle,
+  checkSessionAndRedirect,
```
Comment (di-fabs) [inline src/app.js:11]: It's not a GTM middleware function. You have to put it in another file.

File: src/config/gtmMiddleware.js [hunk 4]
```diff
@@ -45,9 +45,40 @@ const setPageTitle = (req, res, next) => {
 
   next();
 };
+
+const checkSessionAndRedirect = (req, res, next) => {
+  console.log("Middleware executed");
```
Comment (di-fabs) [inline src/config/gtmMiddleware.js:50]: can you remove this console log?

File: src/config/gtmMiddleware.js [hunk 5]
```diff
@@ -45,9 +45,40 @@ const setPageTitle = (req, res, next) => {
...
+  // Check if the user has an active session
+  const hasSession =
+    req.session &&
+    req.session.userSession &&
+    req.session.userSession.startedJourney;
+
+  console.log("User Session:", hasSession);
```
Comment (di-fabs) [inline src/config/gtmMiddleware.js:57]: can you remove this console log?

**Review by di-fabs — CHANGES_REQUESTED** [review]

File: src/config/gtmMiddleware.js (hunk 4 above)
Comment (SophiaDWokoma) [reply src/config/gtmMiddleware.js:50]: done

File: src/app.js (hunk 3 above)
Comment (SophiaDWokoma) [reply src/app.js]: done

File: src/app.js [hunk 6]
```diff
@@ -32,17 +33,26 @@ app.use(
   "/ga4-assets",
   express.static(path.join(__dirname, "../node_modules/one-login-ga4/lib"))
 );
+app.use(
+  session({
+    secret: "your-secret-key", // Should I make this more secure?
```
Comment (Joanneyoung01) [inline src/app.js:38]: Should we maybe add a UUID?

File: src/views/journeyGuard.njk [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{% extends 'common/base.njk' %}
+{% set title = 'Journey Guard' %}
+{%block content%}
+    <h1 class="govuk-heading-l">Sorry, you can not access OneLogin GOV.UK from this page </h1>
```
Comment (Joanneyoung01) [inline src/views/journeyGuard.njk:4]: This looks great

**Review by Joanneyoung01 — APPROVED** [review]

**Pull request merged**